### PR TITLE
Problem (CRO-642): no "watch-only" wallet mode in client

### PR DESCRIPTION
**Pull request by yihuang**
Solution:
- Add these apis to client-rpc:
  - restoreBasic(WalletRequest, String[hex encoded private view key])
  - createStakingWatchAddress(WalletRequest, PublicKey)
  - createTransferWatchAddress(WalletRequest, PublicKey)
  - listUTxO(WalletRequest)
- Add several paramters to transactions api:
  - transactions(WalletRequest, offset: usize, limit: usize, reversed: bool)
- Makes several api changes for testing convinience:
  - Add private option to `view_key` to get private view key.
  - Add `list_pub_key` command to client-cli to list public keys of
transfer addresses.
  - Also record the public key in wallet when creating transfer address.

Implemented integration tests in python.

### CODE REVIEW

**Comment by tomtau** [comment]
There's now a strict mode that fails builds on warnings (e.g. unused import)

**Comment by tomtau** [comment]
bors try

**Comment by tomtau** [comment]
One thing to check: are we using the hardened derivation for HD-wallet? otherwise by knowing the view key, one can get the rest

File: client-rpc/src/rpc/wallet_rpc.rs [hunk 1]
```diff
@@ -132,6 +158,16 @@ where
         Ok(request.name)
     }
 
+    fn restore_watch(&self, request: WalletRequest, view_key: String) -> Result<String> {
```
Comment (tomtau) [inline client-rpc/src/rpc/wallet_rpc.rs]: `view_key: SecUtf8`?

File: integration-tests/bot/tests/test_watch_only_wallet.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import time
+from chainrpc import RPC
+from uuid import uuid1
+
+rpc = RPC()
+
+
+def test_watch_only_wallet(addresses):
```
Comment (tomtau) [inline integration-tests/bot/tests/test_watch_only_wallet.py]: could this test be easily added to one of the integration test pipelines (either or Travis or Drone)? If not, we can do it in another PR

File: client-rpc/src/rpc/wallet_rpc.rs (hunk 1 above)
Comment (yihuang) [reply client-rpc/src/rpc/wallet_rpc.rs]: > One thing to check: are we using the hardened derivation for HD-wallet? otherwise by knowing the view key, one can get the rest

I guess view key should be generated as a leaf, not the root key.

File: integration-tests/bot/tests/test_watch_only_wallet.py (hunk 2 above)
Comment (yihuang) [reply integration-tests/bot/tests/test_watch_only_wallet.py]: > could this test be easily added to one of the integration test pipelines (either or Travis or Drone)? If not, we can do it in another PR

I'll consult @leejw51

File: client-rpc/src/rpc/wallet_rpc.rs (hunk 1 above)
Comment (yihuang) [reply client-rpc/src/rpc/wallet_rpc.rs]: > `view_key: SecUtf8`?

Done

Comment (tomtau) [reply client-rpc/src/rpc/wallet_rpc.rs]: > It's generated the same as a transfer address

It depends on whether it's hardened: https://bitcoin.stackexchange.com/questions/56916/derivation-of-parent-private-key-from-non-hardened-child

Comment (yihuang) [reply client-rpc/src/rpc/wallet_rpc.rs]: > > It's generated the same as a transfer address
> 
> It depends on whether it's hardened: https://bitcoin.stackexchange.com/questions/56916/derivation-of-parent-private-key-from-non-hardened-child

Oh, I see, it's not hardened at the lowest level. Use a different account index for view key would ok then.

Comment (tomtau) [reply client-rpc/src/rpc/wallet_rpc.rs]: not sure if this breaks anything -- @leejw51 @leejw51crypto @devashishdxt -- but I guess it'll be good to change, or add a warning fixme note

Comment (devashishdxt) [reply client-rpc/src/rpc/wallet_rpc.rs]: It may break a few test cases. Also, the generated addresses will not be backwards compatible. But, I think we shouldn't worry about backwards compatibility right now because HD wallets were not a part of our last release.

File: client-core/src/service/wallet_service.rs [hunk 3]
```diff
@@ -23,16 +23,19 @@ pub struct Wallet {
     pub staking_keys: BTreeSet<PublicKey>,
     /// root hashes of multi-sig transfer addresses
     pub root_hashes: BTreeSet<H256>,
+    /// is this wallet watch only
+    pub watch_only: bool,
```
[the remark this answers is not in the document]
Comment (yihuang) [reply client-core/src/service/wallet_service.rs]: I think `WalletRole` is a name too broad for this. Actually, I'm thinking of eliminating this boolean variable, because it doesn't really make much difference. A watch-only wallet is actually a fully functioning `BasicWallet` since it already has view key pair.

[the remark this answers is not in the document]
Comment (yihuang) [reply client-core/src/service/wallet_service.rs]: Changed `restoreWatch` to `restoreBasic`, removed the `watch_only` field in `Wallet`.

File: integration-tests/bot/tests/test_watch_only_wallet.py (hunk 2 above)
Comment (yihuang) [reply integration-tests/bot/tests/test_watch_only_wallet.py]: I've added it to the ci script, but can't test locally because of the docker image building failure, maybe try my luck on the CI server.

File: client-rpc/src/rpc/wallet_rpc.rs (hunk 1 above)
Comment (yihuang) [reply client-rpc/src/rpc/wallet_rpc.rs]: I'll implement this in a separate PR then.

**Comment by tomtau** [comment]
bors try

File: integration-tests/bot/tests/test_watch_only_wallet.py (hunk 2 above)
Comment (yihuang) [reply integration-tests/bot/tests/test_watch_only_wallet.py]: It seems no luck with this, I'll revert this now, maybe try again another time.
